src/dos/dos_memory.cpp

```diff
@@ -504,33 +513,56 @@ void DOS_SetupMemory(void) {
 	mcb_sizes+=17;
 	tempmcb2.SetType(middle_mcb_type);
 
-	DOS_MCB mcb((uint16_t)DOS_MEM_START+mcb_sizes);
-	mcb.SetPSPSeg(MCB_FREE);						//Free
-	mcb.SetType(ending_mcb_type); // Last Block
 	if (machine==MCH_TANDY) {
 		/* memory up to 608k available, the rest (to 640k) is used by
 			the tandy graphics system's variable mapping of 0xb800 */
-		mcb.SetSize(0x9BFF - DOS_MEM_START - mcb_sizes);
+		DOS_MCB free_block((uint16_t)DOS_MEM_START+mcb_sizes);
+		free_block.SetPSPSeg(MCB_FREE);
+		free_block.SetType(ending_mcb_type);
+		free_block.SetSize(0x9BFF - DOS_MEM_START - mcb_sizes);
 	} else if (machine==MCH_PCJR) {
-		/* memory from 128k to 640k is available */
-		mcb_devicedummy.SetPt((uint16_t)0x2000);
-		mcb_devicedummy.SetPSPSeg(MCB_FREE);
-		mcb_devicedummy.SetSize(umb_start_seg - 0x2000);
-		mcb_devicedummy.SetType(ending_mcb_type);
-
-		/* exclude PCJr graphics region */
-		mcb_devicedummy.SetPt((uint16_t)0x17ff);
-		mcb_devicedummy.SetPSPSeg(MCB_DOS);
-		mcb_devicedummy.SetSize(0x800);
-		mcb_devicedummy.SetType(middle_mcb_type);
-
-		/* memory below 96k */
-		mcb.SetSize(0x1800 - DOS_MEM_START - (2+mcb_sizes));
-		mcb.SetType(middle_mcb_type);
+		const auto pcjr_start = DOS_MEM_START + mcb_sizes;
+
+		// PCjr video memory uses shared RAM at linear address 96KB - 128KB
+		constexpr auto video_memory_start = kilobytes_to_segment(96);
+
+		const Section_prop* section = static_cast<Section_prop*>(control->GetSection("dosbox"));
+		assert(section);
+		if (section->Get_bool("pcjr_expanded_memory")) {
+			// With expanded memory. reserve the lower memory up to video memory
+			// This makes application memory contiguous in order to prevent crashes
+			// This is needed to prevent Sierra AGI games from crashing
```

Comment:
Minor: fix grammar (comma after "with expanded memory", periods at the ends of the sentences).

Reply:
I would also put a link to this nice article you found here: 

```
// Further details:
// https://www.atarimagazines.com/compute/issue58/pcjr_memory.html
```

Reply:
Fixed